ogr/ogrsf_frmts/generic/ogrlayer.cpp

```diff
@@ -227,6 +227,85 @@ OGRErr OGRLayer::GetExtent(int iGeomField, OGREnvelope *psExtent, int bForce)
 }
 
+OGRErr OGRLayer::GetExtent3D(int iGeomField, OGREnvelope3D *psExtent3D,
+                             int bForce)
+
+{
+    psExtent3D->MinX = 0.0;
+    psExtent3D->MaxX = 0.0;
+    psExtent3D->MinY = 0.0;
+    psExtent3D->MaxY = 0.0;
+    psExtent3D->MinZ = 0.0;
+    psExtent3D->MaxZ = 0.0;
+
+    /* -------------------------------------------------------------------- */
+    /*      If this layer has a none geometry type, then we can             */
+    /*      reasonably assume there are not extents available.              */
+    /* -------------------------------------------------------------------- */
+    if (iGeomField < 0 || iGeomField >= GetLayerDefn()->GetGeomFieldCount() ||
+        GetLayerDefn()->GetGeomFieldDefn(iGeomField)->GetType() == wkbNone)
+    {
+        if (iGeomField != 0)
+        {
+            CPLError(CE_Failure, CPLE_AppDefined,
+                     "Invalid geometry field index : %d", iGeomField);
+        }
+        return OGRERR_FAILURE;
+    }
+
+    /* -------------------------------------------------------------------- */
+    /*      If not forced, we should avoid having to scan all the           */
+    /*      features and just return a failure.                             */
+    /* -------------------------------------------------------------------- */
+    if (!bForce)
+        return OGRERR_FAILURE;
+
+    /* -------------------------------------------------------------------- */
+    /*      OK, we hate to do this, but go ahead and read through all       */
+    /*      the features to collect geometries and build extents.           */
+    /* -------------------------------------------------------------------- */
+    OGREnvelope3D oEnv;
+    bool bExtentSet = false;
+
+    for (auto &&poFeature : *this)
+    {
+        OGRGeometry *poGeom = poFeature->GetGeomFieldRef(iGeomField);
+        if (poGeom == nullptr || poGeom->IsEmpty())
+        {
+            /* Do nothing */
+        }
+        else if (!bExtentSet)
+        {
+            poGeom->getEnvelope(psExtent3D);
+            if (!(CPLIsNan(psExtent3D->MinX) || CPLIsNan(psExtent3D->MinY) ||
+                  CPLIsNan(psExtent3D->MaxX) || CPLIsNan(psExtent3D->MaxY)) ||
+                CPLIsNan(psExtent3D->MinZ) || CPLIsNan(psExtent3D->MaxZ))
+            {
+                bExtentSet = true;
+            }
+        }
+        else
+        {
+            poGeom->getEnvelope(&oEnv);
+            if (oEnv.MinX < psExtent3D->MinX)
+                psExtent3D->MinX = oEnv.MinX;
+            if (oEnv.MinY < psExtent3D->MinY)
+                psExtent3D->MinY = oEnv.MinY;
+            if (oEnv.MinZ < psExtent3D->MinZ)
+                psExtent3D->MinZ = oEnv.MinZ;
+            if (oEnv.MaxX > psExtent3D->MaxX)
+                psExtent3D->MaxX = oEnv.MaxX;
+            if (oEnv.MaxY > psExtent3D->MaxY)
+                psExtent3D->MaxY = oEnv.MaxY;
+            if (oEnv.MaxZ > psExtent3D->MaxZ)
+                psExtent3D->MaxZ = oEnv.MaxZ;
```

Comment:
You might want to use the OGREnvelope3D::Merge() method. It might perhaps need tweeking to deal with NaN in Z components

Reply:
I am trying to handle NaN but before going down the rabbit hole of patching all getEnvelope() implementations (and the OGR geometry classes which initialize Z to 0) I wanted to ask you if it worth the effort.

Maybe it is better to check if geom is 3D and bail out from GetExtent3D if it's not.

Reply:
@rouault I am trying a mixed approach (keep NaN for the envelope but not dig into the geometry classes) see https://github.com/OSGeo/gdal/pull/8806/commits/9783464d62429d9df209289c4a9c80c049e6bd05#diff-65b11183cce13d2ef546cfbe7935e31cc23b6557aae65276d8b8494a3a80720fR2939 for the new behavior, does it make sense?

Reply:
> I am trying a mixed approach (keep NaN for the envelope but not dig into the geometry classes) see [9783464#diff-65b11183cce13d2ef546cfbe7935e31cc23b6557aae65276d8b8494a3a80720fR2939](https://github.com/OSGeo/gdal/commit/9783464d62429d9df209289c4a9c80c049e6bd05#diff-65b11183cce13d2ef546cfbe7935e31cc23b6557aae65276d8b8494a3a80720fR2939) for the new behavior, does it make sense?

yes, that's doable. Updating OGRGeometry::getEnvelope() to set NaN for Z could also be done as a later independent step.